### PR TITLE
Use given matched resources when building Mesos protobufs for pods

### PR DESCRIPTION
**Pull request by ichernetsky**
Previously, passed resources were ignored, and we just constructed new
resource protobufs based on a pod definition.

JIRA issues: [MARATHON-8110](https://jira.mesosphere.com/browse/MARATHON-8110)

### CODE REVIEW

**Comment by ichernetsky** [comment]
I ran `test_pod_with_persistent_volume_recovers` in a loop for ~12 hours. No single failure.

File: src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala [hunk 1]
```diff
@@ -83,6 +86,31 @@ object TaskGroupBuilder extends StrictLogging {
...
   }
 
+  // This method just double-checks that the matched resources are exactly
+  // what is needed and in right quantities.
+  @SuppressWarnings(Array("ComparingFloatingPointTypes"))
+  private[this] def verifyResources(pod: PodDefinition, matchedResources: Seq[mesos.Resource]): Unit = {
+    assert(matchedResources.count(_.getName == "cpus") <= 1)
+    assert(matchedResources.count(_.getName == "mem") <= 1)
+    assert(matchedResources.count(_.getName == "disk") <= 1)
+    assert(matchedResources.count(_.getName == "gpus") <= 1)
```
Comment (jeschkies) [inline src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: I wonder if we could define resources that it's impossible to have multiple definitions. I'm having a similar issue with the Rust client for Mesos.

File: src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala [hunk 2]
```diff
@@ -43,13 +43,16 @@ object TaskGroupBuilder extends StrictLogging {
...
+    verifyResources(podDefinition, resourceMatch.resources)
+
     val allEndpoints = podDefinition.containers.flatMap(_.endpoints)
 
     val mesosNetworks = buildMesosNetworks(
       podDefinition.networks, allEndpoints, resourceMatch.hostPorts, config.mesosBridgeName)
 
     val executorInfo = computeExecutorInfo(
       podDefinition,
+      resourceMatch.resources,
```
Comment (jeschkies) [inline src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: Ah, so this is the fix?

Comment (ichernetsky) [reply src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: The fix is how `scalarResource` is changed and how it is used now.

File: src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala (hunk 1 above)
Comment (ichernetsky) [reply src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: I don't really know :(

Comment (timcharper) [reply src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: In the case of disk resources, it is definitely possible to have multiple matched resources (there are different types of disk resources: root, path, mount)

Comment (ichernetsky) [reply src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: This is not about persistent volumes.

Comment (ichernetsky) [reply src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: I talked to Greg about it, and you are right. If more disks are specified using `--resources` flag, then Mesos can send an offer with two or more disks.

Comment (timcharper) [reply src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: Could this code be run against a launch attempt for a pod with persistent volumes? Sorry, probably lacking some context here, but I don't fully see how persistent volumes aren't potentially a factor here.

Comment (ichernetsky) [reply src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: Persistent volume resources are passed separately.
```
    volumeMatchOption: Option[PersistentVolumeMatcher.VolumeMatch]
```

**Comment by ichernetsky** [comment]
Rebased.

**Comment by ichernetsky** [comment]
An unrelated failure.

File: src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala [hunk 3]
```diff
@@ -83,6 +86,189 @@ object TaskGroupBuilder extends StrictLogging {
...
+      }
+    }
+
+    private def consume(
+      resources: List[(BigDecimal, mesos.Resource)],
+      quantity: BigDecimal): Option[(List[(BigDecimal, mesos.Resource)], mesos.Resource)] =
+      if (quantity <= BigDecimal(0)) {
+        None
+      } else {
+        resources.partition(_._1 != quantity) match {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: What is `_._1`? How about making this explicit?

File: src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala [hunk 4]
```diff
@@ -83,6 +86,189 @@ object TaskGroupBuilder extends StrictLogging {
...
+    // Find a GPU resource for the given quantity if possible, and remove the resource from the resource list.
+    def consumeGpus(quantity: Double): Option[mesos.Resource] = {
+      consume(gpus, BigDecimal(quantity)).map {
+        case (resources, resource) =>
+          gpus = resources
+          resource
+      }
+    }
+
+    private def consume(
```
Comment (jeschkies) [inline src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: I'm not really sure what this method is doing.

File: src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala [hunk 5]
```diff
@@ -83,6 +86,189 @@ object TaskGroupBuilder extends StrictLogging {
...
+    private var cpus = initial.cpus
+    private var mem = initial.mem
+    private var disk = initial.disk
+    private var gpus = initial.gpus
+
+    // Find a CPU resource for the given quantity if possible, and remove the resource from the resource list.
+    def consumeCpus(quantity: Double): Option[mesos.Resource] = {
+      consume(cpus, BigDecimal(quantity)).map {
+        case (resources, resource) =>
+          cpus = resources
```
Comment (jeschkies) [inline src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: This is strange to me. Why are we modifying the state here? How about calling `resource` `consumed_resource`.

File: src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala [hunk 6]
```diff
@@ -83,6 +86,189 @@ object TaskGroupBuilder extends StrictLogging {
...
+            Some((left ::: right, pair._2))
+        }
+      }
+  }
+
+  private[this] def binPackResources(pod: PodDefinition, matchedResources: Seq[mesos.Resource]): Resources = {
+    verifyMatchedResources(pod, matchedResources)
+
+    val decimalZero = BigDecimal(0)
+    val cpuQuantities = {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: Are these the required quantities?

File: src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala [hunk 7]
```diff
@@ -83,6 +86,189 @@ object TaskGroupBuilder extends StrictLogging {
...
+          case (_, Nil) =>
+            throw new IllegalStateException("failed to find a resource with the given quantity")
+          case (left, pair :: right) =>
+            // return a new resource list not containing the found resource and the found resource itself
+            Some((left ::: right, pair._2))
+        }
+      }
+  }
+
+  private[this] def binPackResources(pod: PodDefinition, matchedResources: Seq[mesos.Resource]): Resources = {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: Are the resources from one offer only? Why is it a sequence?

File: src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala (hunk 5 above)
Comment (ichernetsky) [reply src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: I have renamed both variables. Why does it look strange to you?

File: src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala (hunk 4 above)
Comment (ichernetsky) [reply src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: I have added a comment.

File: src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala (hunk 3 above)
Comment (ichernetsky) [reply src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: Done.

File: src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala (hunk 7 above)
Comment (ichernetsky) [reply src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: Mesos folks said that an offer can have multiple resources of one kind (say, `disk`).

File: src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala (hunk 6 above)
Comment (ichernetsky) [reply src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: These are aggregated defined/requested quantities in a pod definition.

File: src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala (hunk 7 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala]: Ah, thanks for clarifying.

**Pull request merged**